SNPipeline.VCF2AB/SNPipeline_VCF2AB.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 2/26/14               \" DATE 
.Dt SNPipeline.VCF2AB 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm SNPipeline.VCF2AB,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

SNPipeline.VCF2AB/SNPipeline_VCF2AB.cpp
/*
 * Copyright (c) 2016 Hailin Su, ISU NBCEC
 *
 * This file is part of SNPipeline.
 *
 * SNPipeline is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SNPipeline is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public License
 * along with SNPipeline.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  SNPipeline_VCF2AB.cpp
//  SNPipeline
//
//  Created by Hailin SU on 2/26/14.
//

#include "SNPipeline_VCF2AB.h"
#include <iomanip>

#define I(...)              lg.generalLog("I", __VA_ARGS__)
#define E(str, ...)         lg.generalLog("E", str, 1, __VA_ARGS__)
#define F(str, ...)         lg.generalLog("F", str, 1, __VA_ARGS__)
#define L(...)              if(vb) lg.generalLog("I", __VA_ARGS__)

using namespace std;

namespace libcbk
{
    void SNPipeline_VCF2AB::usage(const string& argv0)
    {
        cout << endl
        << "     version: " << version << endl
        << "####################################################################" << endl
        << "#######" << endl
        << "###                 " << progName << endl
        << "####" << endl
        << "############" << endl
        << "##                              by hailins@iastate.edu" << endl
        << "#####                           created " << created << endl
        << "####                            updated " << updated << endl
        << "#######" << endl
        << "####################################################################" << endl
        << endl << "    " << descrip << endl
        << endl
        << "usage:" << endl
        << argv0 << " VCF4.1_File" << endl << endl
        << endl     << endl;
    }

    SNPipeline_VCF2AB::SNPipeline_VCF2AB(const int argc, const char * argv[])
    {
        progName    = "SNPipeline.VCF2AB";
        version     = "2014-02-26 11:33 CDT";
        created     = "2014-02-26";
        updated     = "N/A";
        descrip     = "SNPipeline - setp ?. Convert phased/unphased, but must be imputed VCF4.1 file to merged/unmerged AB genotype format.";
        //        displayInfo();
        lg.initLogger();
        theMain(argc, argv);
        lg.finalizeLogger();
        /*
         updates
         2014-02-11 11:33 AM init
         */
    }

    void SNPipeline_VCF2AB::theMain(const int argc, const char * argv[])
    {
        // **************** parsing parameters
        if(argc!=2)
        {
            usage(argv[0]);
            exit (-1);
        }
        //string vcfFileName (argv[1]);
        VCF_File v(argv[1]);

        if(!v.isValid)
            return;

        //                           0     1     2     3
        //        const string vecUC2STR[4] {"--", "AA", "BB", "AB"};
        const string vecUC2STR[] = {"--", "AA", "BB", "AB"};

        const string outFile ("ab-genotype.from." + v.fileName + ".abg");

        ofstream ofs (outFile.c_str());

        if(!ofs)
        {
            E("s", "ERROR: cannot open file to write into.");
            exit (-30);
        }

        // TODO: write out the ab-genotype file, from SNPipeline/data.cpp
        unsigned long long i, j;
        const unsigned long numberSamples (v.vecSampleIDs.size());
        const unsigned long numberSNPs (v.vecSNPNames.size());
        bool vb(true);

        L("s", 1, "Start unpacking genotypes and writing them out...");

        // write out the SNP names for the whole file
        ofs << "#SNP_IDs";
        for(i=0; i<numberSNPs; ++i)
            ofs << "#" << v.vecSNPNames[i];
        ofs << endl;

        unsigned long long segIndex(0);       // each char for 4 segments [ 11 01 00 10 ]
        const unsigned long long numx(numberSNPs*numberSamples);
        unsigned char shift;
        for(i=0; i<numberSamples; ++i)
        {
            if(i%7==0)  // show % every 7 samples
                progShow('W', unsigned(100*i/numberSamples));

            ofs << left << setw(32) << v.vecSampleIDs[i];
            for(j=0; j<numx; j+=numberSamples)
            {
                //              segIndex =  j * numberSamples   +   i;
                //          [base]   [offset]
                segIndex =  j    +   i;
                //                charIndex = segIndex / 4;           // the index of vecGenotype for the current Animal
                /*
                 // determine the position of genotype for the current animal
                 //     2bit  3*2 0s in the left    curSeg
                 shift= 2 *   (3                  - segIndex % 4);

                 mask =  3 << shift;    //  0000 0011 shift left to get the mask

                 ctmp = vecGenotypes[charIndex];
                 ctmp &= mask;
                 // shift ctmp right back to get the coded genotype
                 ctmp >>= shift;

                 // at last,
                 //          mask          char = 4-seg-genotypes        shift
                 //ctmp = ( (3 << shift) & vecGenotypes[charIndex] ) >> (2 * (3 - segIndex % 4));
                 ofs << vecUC2STR[ctmp] << " ";

                 *****************
                 *
                 *   for speed optimization
                 *   detailed algorithm see above
                 */
                shift = 2 * (3 - (segIndex % 4));
                ofs  << " " << vecUC2STR[( (3 << shift) & v.vecGenotypes[segIndex/4] ) >> shift];
            }
            ofs << endl;
        }
        progClear();

        ofs.close();
        v.clear();
    }
}

int main(const int argc, const char * argv[])
{
    libcbk::SNPipeline_VCF2AB(argc, argv);
    return 0;
}
